src/trajectory.jl

```diff
@@ -115,90 +117,152 @@ Multinomial sampler carried during the building of the tree.
 It contains the weight of the tree, defined as the total probabilities of the leaves.
 """
 struct MultinomialTreeSampler{F<:AbstractFloat} <: AbstractTreeSampler
+    zcand   ::  PhasePoint
     logw    ::  F     # total energy for the given tree, i.e. sum of energy of all leaves
 end
 
 """
 Slice sampler for the starting single leaf tree.
 Slice variable is initialized.
 """
-SliceTreeSampler(rng::AbstractRNG, H::AbstractFloat) = SliceTreeSampler(log(rand(rng)) - H, 1)
+SliceTreeSampler(rng::AbstractRNG, z0::PhasePoint) = SliceTreeSampler(z0, log(rand(rng)) - energy(z0), 1)
 
 """
 Multinomial sampler for the starting single leaf tree.
-Tree weight is just the probability of the only leave.
+Tree weight is the (unnormalised) energy of the leaf.
+
+Ref: https://github.com/stan-dev/stan/blob/develop/src/stan/mcmc/hmc/nuts/base_nuts.hpp#L226
 """
-MultinomialTreeSampler(rng::AbstractRNG, H::AbstractFloat) = MultinomialTreeSampler(-H)
+MultinomialTreeSampler(rng::AbstractRNG, z0::PhasePoint) = MultinomialTreeSampler(z0, zero(energy(z0)))
 
 """
 Create a slice sampler for a single leaf tree:
 - the slice variable is copied from the passed-in sampler `s` and
 - the number of acceptable candicates is computed by comparing the slice variable against the current energy.
 """
-makebase(s::SliceTreeSampler, H::AbstractFloat) = SliceTreeSampler(s.logu, (s.logu <= -H) ? 1 : 0)
+SliceTreeSampler(s::SliceTreeSampler, H0::AbstractFloat, zcand::PhasePoint) = 
+    SliceTreeSampler(zcand, s.logu, (s.logu <= -energy(zcand)) ? 1 : 0)
 
 """
-Create a multinomial sampler for a single leaf tree:
-- the tree weight is just the probability of the only leave.
+Multinomial sampler for the starting single leaf tree.
+- tree weight is the (unnormalised) energy of the leaf.
 """
-makebase(s::MultinomialTreeSampler, H::AbstractFloat) = MultinomialTreeSampler(-H)
+MultinomialTreeSampler(s::MultinomialTreeSampler, H0::AbstractFloat, zcand::PhasePoint) = 
+    MultinomialTreeSampler(zcand, H0 - energy(zcand))
+
+function combine(rng::AbstractRNG, s1::SliceTreeSampler, s2::SliceTreeSampler)
+    @assert s1.logu == s2.logu "Cannot combine two slice sampler with different slice variable"
+    n = s1.n + s2.n
+    zcand = rand(rng) < s1.n / n ? s1.zcand : s2.zcand
```

Comment:
Perhaps wrap `rand(rng) < s1.n / n ` into a function

```julia
rand_bool(rng::AbstractRNG, prob::T) where {T <: AbstractFloat} = rand(rng, T) ≤ prob
```

since it is used repeatedly in many places.

Reply:
Maybe call `rand(Bernoulli(prob))`?

Reply:
Sounds good, but maybe passing the`rng` explicitly.

Reply:
Sure.